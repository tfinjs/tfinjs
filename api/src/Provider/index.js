import JsToHcl from '../JsToHcl';

/**
 * Creates a new provider instance for an Api
 *
 * @param {string} provider name
 * @param {object} body of the provider config
 * @param {providerUri} providerUri
 * @class Provider
 */
class Provider {
  constructor(provider, body, uri) {
    this.provider = provider;
    this.body = body;
    this.uri = uri;
  }

  /**
   * Get the providerUri
   *
   * @returns {providerUri} providerUri
   * @memberof Provider
   */
  getUri() {
    return this.uri;
  }

  /**
   * Gets the HCL for the provider
   *
   * @returns {hcl} The provider code
   * @memberof Provider
   */
  getHcl() {
    const converter = new JsToHcl();
    const providerBody = converter.stringify(this.body);
    return `provider "${this.provider}" { ${providerBody} }`;
  }
}

export default Provider;
