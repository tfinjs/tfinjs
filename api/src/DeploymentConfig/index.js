import assert from 'assert';
import Namespace from '../Namespace/NamespaceParent';
import Provider from '../Provider';
import DeploymentConfigParent from './DeploymentConfigParent';

/**
 * Creates an instance of DeploymentConfig.
 *
 * @param {namespace} namespace - An instance of the Namespace class
 * @param {object} config - Configuration object for the deployment config
 * @param {string} config.environment - The deployment environment
 * @param {string} config.version - The deployment version
 * @param {provider} config.provider - And instance of the Provider class which defines to which tennant and location resources with this config should be deployed
 * @class DeploymentConfig
 */
class DeploymentConfig extends DeploymentConfigParent {
  constructor(namespace, { environment, version, provider }) {
    super();
    assert(
      namespace instanceof Namespace,
      'namespace must be an instance of Namespace',
    );
    assert(typeof environment === 'string', 'environment must be string');
    assert(typeof version === 'string', 'version must be string');
    assert(
      provider instanceof Provider,
      'provider must be an instance of Provider',
    );

    this.namespace = namespace;
    this.environment = environment;
    this.version = version;
    this.provider = provider;
  }

  /**
   * Gets the uri of the API.
   * Is unique based on
   * project, environment, version,
   * providerUri and the namespace.
   *
   * @returns {apiUri} apiUri - The Api uri
   * @memberof DeploymentConfig
   */
  getUri() {
    const { environment, version, namespace } = this;
    const project = this.namespace.project.getValue();

    assert(
      namespace instanceof Namespace,
      'namespace must be an instance of Namespace',
    );
    assert(typeof environment === 'string', 'environment must be a string');
    assert(typeof version === 'string', 'version must be a string');
    assert(typeof project === 'string', 'project value must be a string');
    const uri = `${project}/${environment}/${version}/${this.provider.getUri()}/${namespace.getValue()}`;
    return uri;
  }
}

export default DeploymentConfig;
