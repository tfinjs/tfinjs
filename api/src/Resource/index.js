import { resolve } from 'path';
import mkdirp from 'mkdirp';
import assert from 'assert';
import camelCase from 'lodash/camelCase';
import set from 'lodash/set';
import get from 'lodash/get';
import has from 'lodash/has';
import JsToHcl, { Provisioner } from '../JsToHcl';
import requiredParam from '../statics/requiredParam';
import throwError from '../statics/throwError';
import DeploymentConfig from '../DeploymentConfig';
import createTerraformStringInterpolation from '../statics/createTerraformStringInterpolation';
import resourceExistsInList from '../statics/resourceExistsInList';
import uuid from '../statics/uuid';
import hclPrettify from '../statics/hclPrettify';
import { TERRAFORM_OUTPUT_PREFIX } from '../constants';
import murmur from '../statics/murmur';

const hooks = ['buildHook', 'serializingHook'];

/**
 * Creates an instance of Resource.
 *
 * @param {deploymentConfig} deploymentConfig - DeploymentConfig instance
 * @param {type} type - Resource type, e.g. aws_iam_role
 * @param {name} name - Resource name, some name for the resource
 * @param {body} body - Resource body, the terraform key value pairs
 * @param {object} options - Options
 * @param {resource[]} options.dependsOn - Resources this resource explicitly depends on
 * @class Resource
 */
class Resource {
  constructor(
    deploymentConfig = requiredParam('deploymentConfig'),
    type = requiredParam('type'),
    name = requiredParam('name'),
    body = requiredParam('body'),
    options,
  ) {
    assert(
      deploymentConfig instanceof DeploymentConfig,
      'deploymentConfig must be an instance of DeploymentConfigin Resource.constructor',
    );
    assert(typeof type === 'string', 'type must be string');
    assert(typeof name === 'string', 'name must be string');
    assert(typeof body === 'object', 'name must be object');
    assert(
      typeof options === 'undefined' || typeof options === 'object',
      'options must be an object',
    );
    if (options && options.dependsOn) {
      assert(
        typeof options.dependsOn === 'undefined'
          || Array.isArray(options.dependsOn),
        'options.dependsOn must be an array',
      );
    }

    const makeAddHook = (hookName) => (hook) => {
      const id = uuid();
      this[hookName][id] = hook;
      return id;
    };
    const makeRemoveHook = (hookName) => (id) => {
      delete this[hookName][id];
      return id;
    };

    hooks.forEach((hookName) => {
      const preHookName = camelCase(`pre_${hookName}s`);
      const postHookName = camelCase(`post_${hookName}s`);
      this[preHookName] = [];
      this[postHookName] = [];

      const addPreHookName = camelCase(`add_pre_${hookName}`);
      const addPostHookName = camelCase(`add_post_${hookName}`);
      const removePreHookName = camelCase(`remove_pre_${hookName}`);
      const removePostHookName = camelCase(`remove_post_${hookName}`);

      this[addPreHookName] = makeAddHook(preHookName);
      this[addPostHookName] = makeAddHook(postHookName);
      this[removePreHookName] = makeRemoveHook(preHookName);
      this[removePostHookName] = makeRemoveHook(postHookName);
    });

    this.type = type;
    this.name = name;

    this.deploymentConfig = deploymentConfig;

    this.body = this.parseValue(body);

    this.deploymentConfig.namespace.project.addResource(this);

    this.options = options;
  }

  /**
   * Gets all the resources this resource explicitly and implicitly depends on
   *
   * @returns {resource[]} dependencies
   * @memberof Resource
   */
  getDependencies() {
    return [...this.getRemoteStates(), ...this.getDependsOn()];
  }

  /**
   * Gets the resources this resource implicitly depends on
   *
   * @returns {resource[]} remoteStates
   * @memberof Resource
   */
  getRemoteStates() {
    return this.remoteStates;
  }

  /**
   * Gets the resources this resource explicitly depends on
   *
   * @returns {resource[]} dependsOn
   * @memberof Resource
   */
  getDependsOn() {
    if (this.options && this.options.dependsOn) {
      return this.options.dependsOn;
    }
    return [];
  }

  /**
   * Gets backend
   *
   * @returns {backend} backend
   * @memberof Resource
   */
  getBackend() {
    return this.getProject().backend;
  }

  /**
   * Gets project
   *
   * @returns {project} project
   * @memberof Resource
   */
  getProject() {
    return this.getNamespace().project;
  }

  /**
   * Gets namespace
   *
   * @returns {namespace} namespace
   * @memberof Resource
   */
  getNamespace() {
    return this.getDeploymentConfig().namespace;
  }

  /**
   * Gets deploymentConfig
   *
   * @returns {deploymentConfig} deploymentConfig
   * @memberof Resource
   */
  getDeploymentConfig() {
    return this.deploymentConfig;
  }

  /**
   * Gets the resource body
   *
   * @param {string|number=} key
   * @returns {body} body - Resource body or the value of the key
   * @memberof Resource
   */
  getBody(key) {
    if (typeof key === 'string' || typeof key === 'number') {
      return get(this.body, key);
    }
    return {
      ...this.body,
      __tfinjsDeleteProvisioner__: new Provisioner('local-exec', {
        when: 'destroy',
        command:
          "require('@tfinjs/api/utils').saveDeploymentStatus('${path.root}', 'DESTROYED')",
        interpreter: ['node', '-e'],
      }),
    };
  }

  /**
   * Update a key in the hcl body
   *
   * @param {string} key
   * @param {body} val
   * @memberof Resource
   */
  updateBody(key, val) {
    set(this.body, key, this.parseValue(val));
  }

  /**
   * Gets the resource type
   *
   * @returns {type} type - Resource type
   * @memberof Resource
   */
  getType() {
    return this.type;
  }

  /**
   * Gets the resource name
   *
   * @returns {name} name - Resource name
   * @memberof Resource
   */
  getName() {
    return this.name;
  }

  /**
   * Recursivly searches through the params for values of type function and then evaulates that function
   *
   * @param {params} params - params
   * @returns {params} params - params
   * @memberof Resource
   */
  parseValue(params = requiredParam('params')) {
    let result = params;
    if (typeof params === 'function' && !(params instanceof Provisioner)) {
      result = params(this);
    } else if (typeof params === 'object' && !Array.isArray(params)) {
      result = this.mapObject(params);
    } else if (Array.isArray(params)) {
      result = this.mapArray(params);
    }
    if (result === null || typeof result === 'undefined') {
      throwError('Value cannot be null or undefined', this.parseValue);
    }
    return result;
  }

  /**
   * Parses each value in an object
   *
   * @param {params} params - params
   * @returns {params} params - params
   * @memberof Resource
   */
  mapObject(params = requiredParam('params')) {
    return Object.entries(params).reduce((c, [key, value]) => {
      const result = this.parseValue(value);
      return {
        ...c,
        [key]: result,
      };
    }, {});
  }

  /**
   * Parses each value in an array
   *
   * @param {params} params - params
   * @returns {params} params - params
   * @memberof Resource
   */
  mapArray(params = requiredParam('params')) {
    return params.map((value) => this.parseValue(value));
  }

  /**
   * Create the versioned name which is derived from the 7 parametes: project, environment, version, platform, namespace, type and name
   *
   * @returns {versionedName} versionedName - The versioned name
   * @memberof Resource
   */
  versionedName() {
    /* must depend on these 7 parameters */
    const uri = this.getUri();

    /* length is 30: 3 from tij prefix, 20 from projectName and 7 from murmur */
    /* must be under 32 */

    const normalizeProjectName = this.deploymentConfig.namespace.project
      .getValue()
      .replace(/[^A-Za-z0-9]/g, '')
      .slice(0, 20)
      .toLowerCase();

    const versionedName = `tij${normalizeProjectName}${murmur(uri)}`;
    return versionedName;
  }

  /**
   * Returns the resource uri. This consists of the api uri and the resource type and name.
   *
   * @returns {string} uri
   * @memberof Resource
   */
  getUri() {
    const uri = `${this.deploymentConfig.getUri()}/${this.type}/${this.name}`;
    return uri;
  }

  /**
   * List of remote states to add the the HCL file when deriving it for isolated deployment
   *
   * @memberof Resource
   */
  remoteStates = [];

  /**
   * List of outputs to add the the HCL file when deriving it for isolated deployment
   *
   * @memberof Resource
   */
  outputs = [];

  /**
   * Maybe (if it doesn't already exist) adds a resource to a list which will be converted to a remove data state in the HCL during HCL file generation
   *
   * @param {string} resource - resource
   * @memberof Resource
   */
  registerRemoteState(resource = requiredParam('resource')) {
    assert(
      resource instanceof Resource,
      'resource must be a instance of Resource',
    );

    if (!resourceExistsInList(this.remoteStates, resource)) {
      this.remoteStates.push(resource);
    }
  }

  addOutputKey(key = requiredParam('key')) {
    if (typeof key !== 'string') {
      throwError('key must be a string');
    }
    if (!this.outputs.includes(key)) {
      this.outputs.push(key);
    }
  }

  callHooks(key) {
    Object.entries(this[key]).forEach(([id, hook]) => hook(this, id));
  }

  async callAsyncHooks(key) {
    await Promise.all(
      Object.entries(this[key]).map(async ([id, hook]) => hook(this, id)),
    );
  }

  /**
   * Generates the HCL content of the resource (with remote states)
   *
   * @returns {hcl} hcl - hcl
   * @memberof Resource
   */
  getHcl() {
    this.callHooks('preSerializingHooks');

    const converter = new JsToHcl();
    const resourceHcl = `resource "${this.type}" "${
      this.name
    }" {${converter.stringify(this.getBody())}}`;

    const remoteDataSourcesHcl = this.remoteStates
      .map((resource) => {
        const versionedName = resource.versionedName();
        return resource.deploymentConfig.namespace.project.backend.getDataConfig(
          versionedName,
        );
      })
      .join('\n');

    const outputs = this.outputs
      .map(
        (key) =>
          /* tfinjs prefix because of https://github.com/hashicorp/terraform/issues/7982 */
          `output "${TERRAFORM_OUTPUT_PREFIX}${key}" {${converter.stringify({
            value: createTerraformStringInterpolation(
              `${this.type}.${this.name}.${key}`,
            ),
          })}}`,
      )
      .join('\n');

    const providerHcl = this.deploymentConfig.provider.getHcl();
    const backendHcl = this.deploymentConfig.namespace.project.backend.getBackendHcl(
      this.versionedName(),
    );

    const dataHcl = `data "external" "save_latest_deploy" {${converter.stringify(
      {
        depends_on: [`${this.type}.${this.name}`],
        program: [
          'node',
          '-e',
          `require('@tfinjs/api/utils').saveDeploymentStatus('\${path.root}', '${this.versionedName()}')`,
        ],
      },
    )}}`;

    this.callHooks('postSerializingHooks');

    return [
      providerHcl,
      backendHcl,
      dataHcl,
      resourceHcl,
      remoteDataSourcesHcl,
      outputs,
    ].join('\n');
  }

  getOutputFolder() {
    const dist = this.deploymentConfig.namespace.project.getDist();
    const name = this.versionedName();
    const outputFolder = resolve(dist, name);
    return outputFolder;
  }

  async build() {
    await this.callAsyncHooks('preBuildHooks');
    const fs = this.deploymentConfig.namespace.project.getFs();
    const outputFolder = this.getOutputFolder();
    mkdirp.sync(outputFolder, {
      fs,
    });
    const hcl = this.getHcl();

    const prettyHcl = await hclPrettify(hcl);

    fs.writeFileSync(resolve(outputFolder, 'main.tf'), prettyHcl);
    fs.writeFileSync(
      resolve(outputFolder, 'tfinjs.json'),
      JSON.stringify(
        {
          name: this.versionedName(),
          uri: this.getUri(),
          contentHash: this.getContentHash(),
          dependencies: this.getDependencies().map((resource) =>
            resource.getUri()),
        },
        null,
        2,
      ),
    );
    await this.callAsyncHooks('postBuildHooks');
  }

  contentHashSeeds = {
    hcl: (r) => r.getHcl(),
  };

  getContentHashSeeds() {
    return this.contentHashSeeds;
  }

  addContentHashSeed(key, func) {
    assert(
      !has(this.contentHashSeeds, key),
      'You cannot add an already existing seed function',
    );
    const testFunc = func(this);
    assert(
      typeof testFunc === 'string' || typeof testFunc === 'number',
      'func must return number or string',
    );
    assert(
      testFunc === func(this),
      'func must return the samve value each time',
    );
    this.contentHashSeeds[key] = func;
  }

  removeContentHashSeed(key) {
    delete this.contentHashSeeds[key];
  }

  getContentHash() {
    const text = Object.values(this.getContentHashSeeds())
      .map((seedFunc) => seedFunc(this))
      .join('');
    const hash = murmur(text);
    return hash;
  }
}

export default Resource;
