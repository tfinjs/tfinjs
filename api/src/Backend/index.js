import assert from 'assert';
import throwError from '../statics/throwError';
import JsToHcl from '../JsToHcl';
import requiredParam from '../statics/requiredParam';
import Provider from '../Provider';
import DeploymentConfig from '../DeploymentConfig';
import Resource from '../Resource';

/**
 * Creates an instance of Backend.
 *
 * @param {string} backend - the backend name
 * @param {object} config
 * @param {function=} config.backendConfig - Config for the terraform { backend { ... } }
 * @param {function=} config.dataConfig - Config for the terraform_remote_state
 * @param {provider=} config.provider - The provider which will be used to deploy the backend source
 * @param {function=} config.create - A function taking resource creating function as the first argument and expects the resource to be returned. The resource should be the backend source.
 * @class Backend
 */
class Backend {
  constructor(
    backend = requiredParam('backend'),
    {
      backendConfig, dataConfig, provider, create,
    } = {},
  ) {
    assert(
      typeof backend === 'string' || backend === null,
      'backend must be string or null',
    );
    assert(
      typeof backendConfig === 'undefined'
        || typeof backendConfig === 'function',
      'backendConfig must be function if defined',
    );
    assert(
      typeof dataConfig === 'undefined' || typeof dataConfig === 'function',
      'dataConfig must be function if defined',
    );
    assert(
      typeof create === 'function' || typeof create === 'undefined',
      'create must be function if defined',
    );
    assert(
      typeof provider === 'undefined' || provider instanceof Provider,
      'provider must be instance of Provider if defined',
    );

    if ((provider && !create) || (!provider && create)) {
      throwError(
        'You must provide both the provider and the create function',
        this.constructor,
      );
    }

    this.backend = backend;
    this.backendConfig = backendConfig;
    this.dataConfig = dataConfig;

    this.provider = provider;
    this.privateCreate = create;
  }

  /**
   * If the backend should be created, (only if provider and create has been provided)
   *
   * @returns {boolean} shouldCreateBackend - If the backend should be created
   * @memberof Backend
   */
  shouldCreateBackend() {
    return this.provider && this.create;
  }

  /**
   * Gets the provided provider
   *
   * @returns {provider} provider - Instance of provider optionally provided to the constructor
   * @memberof Backend
   */
  getProvider() {
    return this.provider;
  }

  /**
   * Gets the create function which will create the resource that constitutes the backend
   *
   * @returns {function} create - The create function optionally provided to the contructor
   * @memberof Backend
   */
  create(deploymentConfig) {
    assert(
      deploymentConfig instanceof DeploymentConfig,
      'deploymentConfig must be an instance of DeploymentConfig, in Backend.create',
    );
    const result = this.privateCreate(deploymentConfig);
    assert(
      result instanceof Resource,
      'You must return an instance of Resource',
    );
    return result;
  }

  /**
   * Creates the terraform backend configuration in HCL
   *
   * @param {versionedName} versionedName
   * @returns {hcl} hcl
   * @memberof Backend
   */
  getBackendHcl(versionedName = requiredParam('versionedName')) {
    assert(typeof versionedName === 'string', 'versionedName must be string');

    if (this.backend === null) {
      return '';
    }

    assert(
      this.backendConfig,
      'You must define the backendConfig on the backend in order to make other resources be able to reference the resource (unless the specified backend is null)',
    );

    const jsToHcl = new JsToHcl();
    const hcl = `terraform { backend "${this.backend}" { ${jsToHcl.stringify(
      this.backendConfig(versionedName),
    )} } }`;

    return hcl;
  }

  /**
   * Creates the terraform_remote_state configuration in HCL
   *
   * @param {versionedName} versionedName
   * @returns {hcl} hcl
   * @memberof Backend
   */
  getDataConfig(versionedName = requiredParam('versionedName')) {
    assert(typeof versionedName === 'string', 'versionedName must be string');
    assert(
      this.dataConfig,
      'You must define the dataConfig on the backend in order to be able to reference other resources',
    );

    const jsToHcl = new JsToHcl();
    const params = {
      backend: 's3',
      config: this.dataConfig(versionedName),
    };
    const hcl = `data "terraform_remote_state" "${versionedName}" {${jsToHcl.stringify(
      params,
    )}}`;

    return hcl;
  }
}
export default Backend;
