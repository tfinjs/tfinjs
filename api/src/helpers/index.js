import assert from 'assert';
import Resource from '../Resource';
import createTerraformStringInterpolation from '../statics/createTerraformStringInterpolation';
import { TERRAFORM_OUTPUT_PREFIX } from '../constants';

/**
 * Factory function to a function where the returned function takes a resource and returns the versionedName of the resource.
 *
 * @returns {function} versionedName
 */
export const versionedName = () => (resource) => {
  assert(
    resource instanceof Resource,
    'resource must be an instance of Resource',
  );
  return resource.versionedName();
};

/**
 * Factory function to a function which takes a resource and returns the terraform interpolation value which references the key
 *
 * @param {resource} resource - a resource
 * @param {string} key - the key referenced on the resource
 * @returns {function} reference - gets the reference string of the resource
 */
export const reference = (resource, key) => {
  assert(
    resource instanceof Resource,
    'resource must be an instance of Resource',
  );
  assert(typeof key === 'string', 'key must be string');

  return (referencedResource) => {
    assert(
      referencedResource instanceof Resource,
      'referencedResource must be an instance of Resource',
    );
    assert(
      referencedResource !== resource,
      'you cannot reference the resource itself',
    );
    referencedResource.registerRemoteState(resource);
    resource.addOutputKey(key);

    /* prefix because of https://github.com/hashicorp/terraform/issues/7982 */
    return createTerraformStringInterpolation(
      `data.terraform_remote_state.${resource.versionedName()}.${TERRAFORM_OUTPUT_PREFIX}${key}`,
    );
  };
};
