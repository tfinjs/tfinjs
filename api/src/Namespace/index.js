import assert from 'assert';
import Project from '../Project/ProjectParent';
import NamespaceParent from './NamespaceParent';

/**
 * Creates an instance of Namespace.
 *
 * @param {project} project - An instance of the Project class
 * @param {string} namespace - The namespace string
 * @class Namespace
 */
class Namespace extends NamespaceParent {
  constructor(project, namespace) {
    super();
    assert(project instanceof Project);
    assert(typeof namespace === 'string', 'namespace must be a string');

    this.namespace = namespace;
    this.project = project;
  }

  /**
   * Gets the value of the namespace
   *
   * @returns {string} namepsace
   * @memberof Namespace
   */
  getValue() {
    return this.namespace;
  }
}

export default Namespace;
