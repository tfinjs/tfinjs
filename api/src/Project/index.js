import fs from 'fs';
import assert from 'assert';
import { isAbsolute } from 'path';
import Backend from '../Backend';
import requiredParam from '../statics/requiredParam';
import Resource from '../Resource';
import resourceExistsInList from '../statics/resourceExistsInList';
import DeploymentConfig from '../DeploymentConfig';
import Namespace from '../Namespace';
import ProjectParent from './ProjectParent';

/**
 * Creates an instance of Project.
 *
 * @param {string} project - The name of the project
 * @param {backend} backen - An instance of the Backend class
 * @param {string} dist - An absolute path pointing to the destination folder for built resources
 * @param {*} [fileSystem=fs]
 * @class Project
 */
class Project extends ProjectParent {
  constructor(project, backend, dist, fileSystem = fs) {
    super();
    assert(typeof project === 'string', 'project must be a string');
    assert(backend instanceof Backend, 'backend must be instance of Backend');

    if (backend.shouldCreateBackend()) {
      /* create dummy deployment and api where the backend can be created */
      const backendBackend = new Backend(null);
      this.backendProject = new Project(
        project,
        backendBackend,
        dist,
        fileSystem,
      );
      const backendNamespace = new Namespace(
        this.backendProject,
        '__tfinjs__backend__',
      );
      const deploymentConfig = new DeploymentConfig(backendNamespace, {
        environment: '_',
        version: '_',
        provider: backend.getProvider(),
      });
      const backendResource = backend.create(deploymentConfig);
      this.addResource(backendResource);
    }

    this.project = project;
    this.backend = backend;

    this.setFs(fileSystem);
    this.setDist(dist);
  }

  /**
   * Gets the filesystem of this project
   *
   * @returns {*} fs
   * @memberof Project
   */
  getFs() {
    return this.fs;
  }

  /**
   * Sets the filesystem of this project
   *
   * @returns {*} fs
   * @memberof Project
   */
  setFs(fileSystem) {
    assert(
      fileSystem.writeFileSync
        && fileSystem.readFileSync
        && fileSystem.mkdir
        && fileSystem.stat,
      'fileSystem must be implemented as the node fs module',
    );
    if (this.backend.shouldCreateBackend()) {
      this.backendProject.setFs(fileSystem);
    }
    this.fs = fileSystem;
  }

  /**
   * Gets the dist folder of this project
   *
   * @returns {string} dist
   * @memberof Project
   */
  getDist() {
    return this.dist;
  }

  /**
   * Updates the dist folder of this project
   *
   * @param {string} dist
   * @memberof Project
   */
  setDist(dist) {
    assert(
      typeof dist === 'string' && isAbsolute(dist),
      'dist must be a string and an absolute path pointing the the dist folder for tfinjs',
    );
    if (this.backend.shouldCreateBackend()) {
      this.backendProject.setDist(dist);
    }
    this.dist = dist;
  }

  /**
   * Gets the name value of this project
   *
   * @returns {string} project - The name of the project
   * @memberof Project
   */
  getValue() {
    return this.project;
  }

  /**
   * Array of added resources
   *
   * @memberof Deployment
   */
  resources = [];

  /**
   * Registers a new resource
   *
   * @param {resource} resource - The resource to be added
   * @memberof Deployment
   */
  addResource(resource = requiredParam('resource')) {
    assert(
      resource instanceof Resource,
      'resource must be an instance of Resource',
    );
    if (resourceExistsInList(this.resources, resource)) {
      const error = new Error(
        'You currently have multiple resources with the same properties which is not allowed',
      );
      throw error;
    }

    this.resources.push(resource);
  }

  /**
   * Gets the resources of the deployment
   *
   * @returns {resources[]} resources - array of resources
   * @memberof Deployment
   */
  getResources() {
    return this.resources;
  }

  /**
   * Gets the dependency graph of the resources of the project
   *
   * @returns {object} graph
   * @memberof Project
   */
  getDependencyGraph() {
    return this.getResources().reduce(
      (nodes, resource) => ({
        ...nodes,
        [resource.versionedName()]: resource
          .getDependencies()
          .map((dependencyResource) => dependencyResource.versionedName()),
      }),
      {},
    );
  }

  /**
   * Gets a resource of the project based on its uri
   *
   * @param {string} uri
   * @returns {resource} resource
   * @memberof Project
   */
  getResourceFromUri(uri) {
    return this.resources.find((resource) => resource.getUri() === uri);
  }

  /**
   * Gets a resource of the project based on its versioned name
   *
   * @param {string} uri
   * @returns {resource} resource
   * @memberof Project
   */
  getResourceFromVersionedName(versionedName) {
    return this.resources.find(
      (resource) => resource.versionedName() === versionedName,
    );
  }

  async build() {
    await Promise.all(
      this.getResources().map(async (resource) => {
        await resource.build();
      }),
    );
  }
}

export default Project;
