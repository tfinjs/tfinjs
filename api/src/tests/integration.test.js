import { join } from 'path';
import Project from '../Project';
import Backend from '../Backend';
import Provider from '../Provider';
import awsProviderUri from '../Provider/uris/aws';
import Resource from '../Resource';
import Namespace from '../Namespace';
import DeploymentConfig from '../DeploymentConfig';
import { versionedName } from '../helpers';
import saveProject from './saveProject';

/* eslint-env jest */

test('integration', async () => {
  const backendBucketName = 'some-backend-bucket';
  const backendBucketRegion = 'eu-north-1';
  const awsAccoundId = '13371337';
  const dist = join(__dirname, 'io/integration.test.out');

  const backend = new Backend('s3', {
    backendConfig: (name) => ({
      bucket: backendBucketName,
      key: `${name}.terraform.tfstate`,
      region: backendBucketRegion,
    }),
    dataConfig: (name) => ({
      bucket: backendBucketName,
      key: `${name}.terraform.tfstate`,
      region: backendBucketRegion,
    }),
    provider: new Provider(
      'aws',
      {
        region: backendBucketRegion,
        assume_role: {
          role_arn: `arn:aws:iam::${awsAccoundId}:role/DeploymentRole`,
        },
      },
      awsProviderUri(awsAccoundId, backendBucketRegion),
    ),
    create: (deploymentConfig) =>
      new Resource(deploymentConfig, 'aws_s3_bucket', 'terraform_state_prod', {
        bucket: backendBucketName,
        acl: 'private',
        versioning: {
          enabled: true,
        },
      }),
  });

  const project = new Project('pet-shop', backend, dist);

  const namespace = new Namespace(project, 'customers');

  const staticConfig = new DeploymentConfig(namespace, {
    environment: 'prod',
    version: 'static',
    provider: new Provider(
      'aws',
      {
        region: 'eu-north-1',
        assume_role: {
          role_arn: `arn:aws:iam::${awsAccoundId}:role/DeploymentRole`,
        },
      },
      awsProviderUri(awsAccoundId, 'eu-north-1'),
    ),
  });
  const table = new Resource(staticConfig, 'aws_dynamodb_table', 'customers', {
    name: versionedName(),
    read_capacity: 20,
    write_capacity: 20,
    hash_key: 'CustomerId',
  });
  await saveProject(project, join(__dirname, 'io/integration.test.ref'), dist);
});
