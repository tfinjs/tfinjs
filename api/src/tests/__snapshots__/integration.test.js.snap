// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`integration 1`] = `
Object {
  "/tijpetshop19qdr42/main.tf": "provider \\"aws\\" {
  assume_role = {
    role_arn = \\"arn:aws:iam::13371337:role/DeploymentRole\\"
  }

  region = \\"eu-north-1\\"
}

terraform {
  backend \\"s3\\" {
    bucket = \\"some-backend-bucket\\"
    key    = \\"tijpetshop19qdr42.terraform.tfstate\\"
    region = \\"eu-north-1\\"
  }
}

data \\"external\\" \\"save_latest_deploy\\" {
  depends_on = [
    \\"aws_dynamodb_table.customers\\",
  ]

  program = [
    \\"node\\",
    \\"-e\\",
    \\"require('@tfinjs/api').utils.saveDeploymentStatus('\${path.root}', 'tijpetshop19qdr42')\\",
  ]
}

resource \\"aws_dynamodb_table\\" \\"customers\\" {
  hash_key = \\"CustomerId\\"
  name     = \\"tijpetshop19qdr42\\"

  provisioner \\"local-exec\\" {
    command = \\"require('@tfinjs/api').utils.saveDeploymentStatus('\${path.root}', 'DESTROYED')\\"

    interpreter = [
      \\"node\\",
      \\"-e\\",
    ]

    when = \\"destroy\\"
  }

  read_capacity  = 20
  write_capacity = 20
}
",
  "/tijpetshop19qdr42/tfinjs.json": "{
  \\"name\\": \\"tijpetshop19qdr42\\",
  \\"uri\\": \\"pet-shop/prod/static/aws/13371337/eu-north-1/customers/aws_dynamodb_table/customers\\",
  \\"contentHash\\": \\"px3wf3\\",
  \\"dependencies\\": []
}",
  "/tijpetshopjr2h4y/main.tf": "provider \\"aws\\" {
  assume_role = {
    role_arn = \\"arn:aws:iam::13371337:role/DeploymentRole\\"
  }

  region = \\"eu-north-1\\"
}

data \\"external\\" \\"save_latest_deploy\\" {
  depends_on = [
    \\"aws_s3_bucket.terraform_state_prod\\",
  ]

  program = [
    \\"node\\",
    \\"-e\\",
    \\"require('@tfinjs/api').utils.saveDeploymentStatus('\${path.root}', 'tijpetshopjr2h4y')\\",
  ]
}

resource \\"aws_s3_bucket\\" \\"terraform_state_prod\\" {
  acl    = \\"private\\"
  bucket = \\"some-backend-bucket\\"

  provisioner \\"local-exec\\" {
    command = \\"require('@tfinjs/api').utils.saveDeploymentStatus('\${path.root}', 'DESTROYED')\\"

    interpreter = [
      \\"node\\",
      \\"-e\\",
    ]

    when = \\"destroy\\"
  }

  versioning = {
    enabled = true
  }
}
",
  "/tijpetshopjr2h4y/tfinjs.json": "{
  \\"name\\": \\"tijpetshopjr2h4y\\",
  \\"uri\\": \\"pet-shop/_/_/aws/13371337/eu-north-1/__tfinjs__backend__/aws_s3_bucket/terraform_state_prod\\",
  \\"contentHash\\": \\"t2zgrk\\",
  \\"dependencies\\": []
}",
}
`;
